File: src/main/java/bai/deirn/fsn/block/MasterBlock.java
package bai.deirn.fsn.block;

import bai.deirn.fsn.block.entity.MasterBlockEntity;
import net.minecraft.block.Block;
import net.minecraft.block.BlockEntityProvider;
import net.minecraft.block.BlockState;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.text.Text;
import net.minecraft.text.TranslatableText;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.BlockView;
import net.minecraft.world.World;

import java.util.List;

public class MasterBlock extends ModBlock implements BlockEntityProvider {

    public MasterBlock(Settings settings) {
        super(settings);
    }

    @Override
    public BlockEntity createBlockEntity(BlockView view) {
        return new MasterBlockEntity();
    }

    @Override
    public void buildTooltip(ItemStack stack, BlockView view, List<Text> tooltip, TooltipContext options) {
        super.buildTooltip(stack, view, tooltip, options);
        tooltip.add(new TranslatableText(stack.getTranslationKey()+".tooltip2"));
    }

    /*
    @Override
    public BlockState getStateForNeighborUpdate(BlockState state, Direction facing, BlockState neighborState, IWorld world, BlockPos pos, BlockPos neighborPos) {
        Block neighborBlock = neighborState.getBlock();

        if (neighborBlock instanceof ChildBlock) {
            BlockEntity neighborBlockEntity = world.getBlockEntity(neighborPos);
            CompoundTag neighborNbt = neighborBlockEntity.toTag(new CompoundTag());
            boolean neighborHasMaster = neighborNbt.getBoolean("hasMaster");

            if (!neighborHasMaster) {
                CompoundTag masterPos = new CompoundTag();
                masterPos.putInt("x", pos.getX());
                masterPos.putInt("y", pos.getY());
                masterPos.putInt("z", pos.getZ());
                neighborNbt.put("masterPos", masterPos);
                neighborNbt.putBoolean("hasMaster", true);
                neighborBlockEntity.fromTag(neighborNbt);
                neighborBlockEntity.markDirty();
            }
        }

        return state;
    }
     */

    @Override
    public void neighborUpdate(BlockState state, World world, BlockPos pos, Block block, BlockPos neighborPos, boolean moved) {
        super.neighborUpdate(state, world, pos, block, neighborPos, moved);

        BlockState neighborState = world.getBlockState(neighborPos);
        Block neighborBlock = neighborState.getBlock();

        if (neighborBlock instanceof ChildBlock) {
            BlockEntity neighborBlockEntity = world.getBlockEntity(neighborPos);
            CompoundTag neighborNbt = neighborBlockEntity.toTag(new CompoundTag());
            boolean neighborHasMaster = neighborNbt.getBoolean("hasMaster");

            if (!neighborHasMaster) {
                CompoundTag masterPos = new CompoundTag();
                masterPos.putInt("x", pos.getX());
                masterPos.putInt("y", pos.getY());
                masterPos.putInt("z", pos.getZ());
                neighborNbt.put("masterPos", masterPos);
                neighborNbt.putBoolean("hasMaster", true);
                neighborBlockEntity.fromTag(neighborNbt);
                neighborBlockEntity.markDirty();
                world.updateNeighbors(neighborPos, neighborBlock);
            }
        }
    }

}

File: src/main/kotlin/io/gitlab/intended/storagenetworks/block/MasterBlock.kt
package io.gitlab.intended.storagenetworks.block

import io.gitlab.intended.storagenetworks.block.entity.MasterBlockEntity
import io.gitlab.intended.storagenetworks.container.ModContainer
import net.minecraft.block.Block
import net.minecraft.block.BlockEntityProvider
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.nbt.CompoundTag
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockView
import net.minecraft.world.World

class MasterBlock(id: String) : ModBlock(id), BlockEntityProvider {

    override fun createBlockEntity(view: BlockView): BlockEntity = MasterBlockEntity()

    override fun neighborUpdate(
        state: BlockState,
        world: World,
        pos: BlockPos,
        block: Block,
        neighborPos: BlockPos,
        moved: Boolean
    ) {
        super.neighborUpdate(state, world, pos, block, neighborPos, moved)

        val neighborState = world.getBlockState(neighborPos)
        val neighborBlock = neighborState.block

        if (neighborBlock is ChildBlock) {
            val neighborBlockEntity = world.getBlockEntity(neighborPos)
            val neighborNbt = neighborBlockEntity!!.toTag(CompoundTag())
            val neighborHasMaster = neighborNbt.getBoolean("hasMaster")
            if (!neighborHasMaster) {
                val masterPos = CompoundTag()
                masterPos.putInt("x", pos.x)
                masterPos.putInt("y", pos.y)
                masterPos.putInt("z", pos.z)
                neighborNbt.put("masterPos", masterPos)
                neighborNbt.putBoolean("hasMaster", true)
                neighborBlockEntity.fromTag(neighborNbt)
                neighborBlockEntity.markDirty()
                world.updateNeighbors(neighborPos, neighborBlock)
            }
        }
    }

    override fun onUse(
        state: BlockState,
        world: World,
        pos: BlockPos,
        player: PlayerEntity,
        hand: Hand,
        hit: BlockHitResult
    ): ActionResult {
        if (!world.isClient) ModContainer.open(this, player)
        return ActionResult.SUCCESS
    }

}
